diff.go

```diff
@@ -1,155 +1,416 @@
+	diffInfos := []DiffInfo{}
+
+	for _, orderedName := range orderedComponentKeys(oldDef) {
+		oldName := ComponentName(orderedName)
+		oldComponent := oldDef[oldName]
+
+		if newComponent, ok := newDef[oldName]; ok {
+			componentDiffInfos := ComponentDiff(*newComponent, *oldComponent)
+
+			if len(componentDiffInfos) > 0 {
+				diffInfos = append(diffInfos, DiffInfo{
+					Type: DiffInfoComponentUpdated,
+					Old:  oldName.String(),
+					New:  oldName.String(),
+				})
+			}
+		}
+	}
+
+	return diffInfos
+}
+
+// component diff
 
-	for change := range c {
-		changes = append(changes, change)
+func ComponentDiff(oldDef, newDef ComponentDefinition) []DiffInfo {
+	diffInfos := []DiffInfo{}
+
+	diffInfos = append(diffInfos, diffComponentImage(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentEntrypoint(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentPorts(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentEnv(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentVolumes(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentArgs(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentDomains(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentLinks(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentExpose(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentScale(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentPod(oldDef, newDef)...)
+	diffInfos = append(diffInfos, diffComponentSignalReady(oldDef, newDef)...)
+
+	return diffInfos
+}
+
+func diffComponentImage(oldDef, newDef ComponentDefinition) []DiffInfo {
+	diffInfos := []DiffInfo{}
+
+	oldImage := oldDef.Image.String()
+	newImage := newDef.Image.String()
+
+	if oldImage != newImage {
+		diffInfos = append(diffInfos, DiffInfo{
+			Type: DiffInfoComponentImageUpdated,
+			Old:  oldImage,
+			New:  newImage,
+		})
 	}
-	return changes
+
+	return diffInfos
 }
 
-type node interface {
-	Name() string
-	Children() []node
+func diffComponentEntrypoint(oldDef, newDef ComponentDefinition) []DiffInfo {
+	diffInfos := []DiffInfo{}
+
+	if oldDef.EntryPoint != newDef.EntryPoint {
+		diffInfos = append(diffInfos, DiffInfo{
+			Type: DiffInfoComponentEntrypointUpdated,
+			Old:  oldDef.EntryPoint,
+			New:  newDef.EntryPoint,
+		})
+	}
 
-	// Diff is called by diffHierarchy when two nodes have the same name and should be checked
-	// for internal differences. Diffs should be written to changes.
-	Diff(path []string, other node, changes chan<- DiffInfo)
+	return diffInfos
 }
 
-type appNode AppDefinition
+func diffComponentPorts(oldDef, newDef ComponentDefinition) []DiffInfo {
+	diffInfos := []DiffInfo{}
 
-func (n appNode) Name() string { return n.AppName }
-func (n appNode) Children() []node {
-	nodes := []node{}
-	for _, service := range n.Services {
-		nodes = append(nodes, serviceNode(service))
+	oldPorts := oldDef.Ports.String()
+	newPorts := newDef.Ports.String()
```

Comment:
Are you sure the order of ports is always the same?
If I change `80/tcp, 81/tcp` to `81/tcp, 80/tcp` I would not expect to get a change.